protocol/lavasession/provider_session_manager.go

```diff
@@ -157,21 +158,52 @@ func (psm *ProviderSessionManager) GetDataReliabilitySession(address string, epo
 	return singleProviderSession, nil
 }
 
-func (psm *ProviderSessionManager) GetSession(ctx context.Context, address string, epoch uint64, sessionId uint64, relayNumber uint64) (*SingleProviderSession, error) {
+func (psm *ProviderSessionManager) GetSession(ctx context.Context, address string, epoch uint64, sessionId uint64, relayNumber uint64, badgeSession *BadgeSession) (*SingleProviderSession, *ProviderSessionsEpochData, error) {
 	if !psm.IsValidEpoch(epoch) { // fast checking to see if epoch is even relevant
 		utils.LavaFormatError("GetSession", InvalidEpochError, utils.Attribute{Key: "RequestedEpoch", Value: epoch}, utils.Attribute{Key: "blockedEpochHeight", Value: psm.blockedEpochHeight}, utils.Attribute{Key: "blockDistanceForEpochValidity", Value: psm.blockDistanceForEpochValidity})
-		return nil, InvalidEpochError
+		return nil, nil, InvalidEpochError
 	}
 
 	providerSessionsWithConsumer, err := psm.IsActiveConsumer(epoch, address)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
+	}
+
+	var badgeUserEpochData *ProviderSessionsEpochData
+	if badgeSession != nil { // badgeSession
+		exists := psm.getBadgeEpochDataFromProviderSessionWithConsumer(badgeSession.BadgeUser, providerSessionsWithConsumer)
+		if !exists {
+			psm.registerBadgeEpochDataToProviderSessionWithConsumer(badgeSession.BadgeUser, badgeSession.BadgeCuAllocation, providerSessionsWithConsumer)
+		}
+		badgeUserEpochData = psm.getBadgeUserEpochDataPointer(badgeSession.BadgeUser, providerSessionsWithConsumer)
 	}
 
-	return psm.getSingleSessionFromProviderSessionWithConsumer(ctx, providerSessionsWithConsumer, sessionId, epoch, relayNumber)
+	singleSessionFromPSWC, err := psm.getSingleSessionFromProviderSessionWithConsumer(ctx, providerSessionsWithConsumer, sessionId, epoch, relayNumber)
+
+	return singleSessionFromPSWC, badgeUserEpochData, err
+}
+
+func (psm *ProviderSessionManager) getBadgeUserEpochDataPointer(badgeUser string, providerSessionsWithConsumer *ProviderSessionsWithConsumer) (badgeUserEpochData *ProviderSessionsEpochData) {
+	psm.lock.RLock()
+	defer psm.lock.RUnlock()
+	badgeUserEpochData = providerSessionsWithConsumer.badgeEpochData[badgeUser]
```

Comment:
Considering your comment here: https://github.com/lavanet/lava/pull/540#discussion_r1230789856
I have removed this function because of redundancy: 96c77665b69bac242c33cd356382acc1461236bb